{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pathlib\n",
    "import pickle\n",
    "\n",
    "sys.path.append('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-20 10:51:25.181 | WARNING  | pals.reactome:get_neo4j_driver:14 - Using a default neo4j server: bolt://localhost:7687\n",
      "2019-12-20 10:51:25.182 | WARNING  | pals.reactome:get_neo4j_driver:19 - Using a default neo4j username or password: neo4j\n",
      "2019-12-20 10:51:25.227 | WARNING  | pals.reactome:get_neo4j_driver:27 - Failed to connect to graph database: Failed to establish connection to ('127.0.0.1', 7687) (reason [Errno 61] Connection refused)\n",
      "2019-12-20 10:51:25.228 | WARNING  | pals.reactome:<module>:34 - Driver initialisation failed. PALS will run without Reactome support.\n"
     ]
    }
   ],
   "source": [
    "from pals.pimp_tools import get_authentication_token, get_pimp_API_token_from_env, PIMP_HOST, download_from_pimp\n",
    "from pals.feature_extraction import DataSource\n",
    "from pals.pathway_analysis import PALS\n",
    "from pals.common import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate token by logging in to PiMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# username = 'karen' # PiMP username\n",
    "# password = 'temp' # PiMP password\n",
    "# host = 'polyomics.mvls.gla.ac.uk' # server address and port\n",
    "# token = get_authentication_token(host, username, password)\n",
    "# print (token)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assume token is stored in environmental variable *PIMP_API_TOKEN*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "token = get_pimp_API_token_from_env()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_id = 1321 # example beer analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-20 10:51:54.925 | DEBUG    | pals.pimp_tools:download_from_pimp:119 - Trying to load data from temp file: /var/folders/t8/0r47by2j56v3kx2h3nz4bpd40000gn/T/pimp_analysis_1321.p\n",
      "2019-12-20 10:51:54.926 | WARNING  | pals.common:load_obj:91 - Old, invalid or missing pickle in /var/folders/t8/0r47by2j56v3kx2h3nz4bpd40000gn/T/pimp_analysis_1321.p. Please regenerate this file.\n",
      "2019-12-20 10:51:54.927 | DEBUG    | pals.pimp_tools:download_from_pimp:123 - Retrieving data for analysis 1321 from PiMP\n",
      "2019-12-20 10:51:56.619 | DEBUG    | pals.pimp_tools:get_data:33 - http://polyomics.mvls.gla.ac.uk/export/get_ms1_intensities?analysis_id=1321\n",
      "2019-12-20 10:52:16.530 | DEBUG    | pals.pimp_tools:get_data:33 - http://polyomics.mvls.gla.ac.uk/export/get_ms1_peaks?analysis_id=1321\n",
      "2019-12-20 10:52:30.450 | DEBUG    | pals.pimp_tools:get_data:33 - http://polyomics.mvls.gla.ac.uk/export/get_experimental_design?analysis_id=1321\n",
      "2019-12-20 10:52:30.451 | DEBUG    | pals.pimp_tools:download_from_pimp:132 - Caching analysis data for next use\n",
      "2019-12-20 10:52:30.452 | DEBUG    | pals.common:save_obj:76 - Saving <class 'dict'> to /var/folders/t8/0r47by2j56v3kx2h3nz4bpd40000gn/T/pimp_analysis_1321.p\n"
     ]
    }
   ],
   "source": [
    "int_df, annotation_df, experimental_design = download_from_pimp(token, PIMP_HOST, analysis_id, 'kegg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CRC Data Analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "crc_table = pd.read_csv('crc_with_kegg.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Metabolites</th>\n",
       "      <th>Kegg Id</th>\n",
       "      <th>Patient group:CRC</th>\n",
       "      <th>Patient group:CRC.1</th>\n",
       "      <th>Patient group:CRC.2</th>\n",
       "      <th>Patient group:CRC.3</th>\n",
       "      <th>Patient group:CRC.4</th>\n",
       "      <th>Patient group:CRC.5</th>\n",
       "      <th>Patient group:CRC.6</th>\n",
       "      <th>...</th>\n",
       "      <th>Patient group:Polyp.66</th>\n",
       "      <th>Patient group:Polyp.67</th>\n",
       "      <th>Patient group:Polyp.68</th>\n",
       "      <th>Patient group:Polyp.69</th>\n",
       "      <th>Patient group:Polyp.70</th>\n",
       "      <th>Patient group:Polyp.71</th>\n",
       "      <th>Patient group:Polyp.72</th>\n",
       "      <th>Patient group:Polyp.73</th>\n",
       "      <th>Patient group:Polyp.74</th>\n",
       "      <th>Patient group:Polyp.75</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1-Methyladenosine</td>\n",
       "      <td>C02494</td>\n",
       "      <td>390953.9725</td>\n",
       "      <td>567984.4359</td>\n",
       "      <td>558842.7545</td>\n",
       "      <td>476949.0000</td>\n",
       "      <td>398317.0000</td>\n",
       "      <td>411224.0000</td>\n",
       "      <td>412250.6387</td>\n",
       "      <td>...</td>\n",
       "      <td>390493.7447</td>\n",
       "      <td>375952.1076</td>\n",
       "      <td>399195.9585</td>\n",
       "      <td>1.018448e+06</td>\n",
       "      <td>503210.6702</td>\n",
       "      <td>311604.8089</td>\n",
       "      <td>514713.0726</td>\n",
       "      <td>407366.5548</td>\n",
       "      <td>463101.8944</td>\n",
       "      <td>510398.3294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1-Methylhistamine</td>\n",
       "      <td>C05127</td>\n",
       "      <td>34627.3365</td>\n",
       "      <td>52845.9284</td>\n",
       "      <td>31507.3427</td>\n",
       "      <td>29397.0000</td>\n",
       "      <td>38877.0000</td>\n",
       "      <td>25493.0000</td>\n",
       "      <td>44478.1854</td>\n",
       "      <td>...</td>\n",
       "      <td>18266.0890</td>\n",
       "      <td>30752.9780</td>\n",
       "      <td>41086.0892</td>\n",
       "      <td>2.094303e+04</td>\n",
       "      <td>27630.5961</td>\n",
       "      <td>30531.1781</td>\n",
       "      <td>24293.0652</td>\n",
       "      <td>29915.2947</td>\n",
       "      <td>32606.5463</td>\n",
       "      <td>29862.9664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>2-Aminoadipate</td>\n",
       "      <td>C00956</td>\n",
       "      <td>141257.3640</td>\n",
       "      <td>528024.6865</td>\n",
       "      <td>188272.3016</td>\n",
       "      <td>107076.0918</td>\n",
       "      <td>158299.0710</td>\n",
       "      <td>136600.4976</td>\n",
       "      <td>235936.4772</td>\n",
       "      <td>...</td>\n",
       "      <td>149966.9484</td>\n",
       "      <td>188581.2257</td>\n",
       "      <td>153364.1518</td>\n",
       "      <td>2.666709e+06</td>\n",
       "      <td>228074.4199</td>\n",
       "      <td>132828.4060</td>\n",
       "      <td>215838.8101</td>\n",
       "      <td>179582.5152</td>\n",
       "      <td>227973.0218</td>\n",
       "      <td>216277.0586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>2-Deoxyuridine</td>\n",
       "      <td>C00526</td>\n",
       "      <td>13115.8131</td>\n",
       "      <td>14327.6960</td>\n",
       "      <td>12756.8649</td>\n",
       "      <td>10539.0000</td>\n",
       "      <td>11689.0000</td>\n",
       "      <td>12425.0000</td>\n",
       "      <td>15692.6312</td>\n",
       "      <td>...</td>\n",
       "      <td>11790.1935</td>\n",
       "      <td>11041.2359</td>\n",
       "      <td>13050.2462</td>\n",
       "      <td>1.094457e+04</td>\n",
       "      <td>13652.0542</td>\n",
       "      <td>9250.8437</td>\n",
       "      <td>14597.4711</td>\n",
       "      <td>14137.2620</td>\n",
       "      <td>13281.2552</td>\n",
       "      <td>12653.4161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>4-Pyridoxic acid</td>\n",
       "      <td>C00847</td>\n",
       "      <td>333815.0605</td>\n",
       "      <td>402460.7838</td>\n",
       "      <td>389636.3191</td>\n",
       "      <td>373058.7101</td>\n",
       "      <td>323706.9435</td>\n",
       "      <td>358042.2763</td>\n",
       "      <td>355881.3356</td>\n",
       "      <td>...</td>\n",
       "      <td>379298.6241</td>\n",
       "      <td>356057.1511</td>\n",
       "      <td>412505.9735</td>\n",
       "      <td>4.260027e+05</td>\n",
       "      <td>358592.6192</td>\n",
       "      <td>314369.6151</td>\n",
       "      <td>327164.9675</td>\n",
       "      <td>361262.1060</td>\n",
       "      <td>376591.8864</td>\n",
       "      <td>368266.5694</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 227 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        Metabolites Kegg Id  Patient group:CRC  \\\n",
       "0           1  1-Methyladenosine  C02494        390953.9725   \n",
       "1           2  1-Methylhistamine  C05127         34627.3365   \n",
       "2           3     2-Aminoadipate  C00956        141257.3640   \n",
       "3           4     2-Deoxyuridine  C00526         13115.8131   \n",
       "4           5   4-Pyridoxic acid  C00847        333815.0605   \n",
       "\n",
       "   Patient group:CRC.1  Patient group:CRC.2  Patient group:CRC.3  \\\n",
       "0          567984.4359          558842.7545          476949.0000   \n",
       "1           52845.9284           31507.3427           29397.0000   \n",
       "2          528024.6865          188272.3016          107076.0918   \n",
       "3           14327.6960           12756.8649           10539.0000   \n",
       "4          402460.7838          389636.3191          373058.7101   \n",
       "\n",
       "   Patient group:CRC.4  Patient group:CRC.5  Patient group:CRC.6  ...  \\\n",
       "0          398317.0000          411224.0000          412250.6387  ...   \n",
       "1           38877.0000           25493.0000           44478.1854  ...   \n",
       "2          158299.0710          136600.4976          235936.4772  ...   \n",
       "3           11689.0000           12425.0000           15692.6312  ...   \n",
       "4          323706.9435          358042.2763          355881.3356  ...   \n",
       "\n",
       "   Patient group:Polyp.66  Patient group:Polyp.67  Patient group:Polyp.68  \\\n",
       "0             390493.7447             375952.1076             399195.9585   \n",
       "1              18266.0890              30752.9780              41086.0892   \n",
       "2             149966.9484             188581.2257             153364.1518   \n",
       "3              11790.1935              11041.2359              13050.2462   \n",
       "4             379298.6241             356057.1511             412505.9735   \n",
       "\n",
       "   Patient group:Polyp.69  Patient group:Polyp.70  Patient group:Polyp.71  \\\n",
       "0            1.018448e+06             503210.6702             311604.8089   \n",
       "1            2.094303e+04              27630.5961              30531.1781   \n",
       "2            2.666709e+06             228074.4199             132828.4060   \n",
       "3            1.094457e+04              13652.0542               9250.8437   \n",
       "4            4.260027e+05             358592.6192             314369.6151   \n",
       "\n",
       "   Patient group:Polyp.72  Patient group:Polyp.73  Patient group:Polyp.74  \\\n",
       "0             514713.0726             407366.5548             463101.8944   \n",
       "1              24293.0652              29915.2947              32606.5463   \n",
       "2             215838.8101             179582.5152             227973.0218   \n",
       "3              14597.4711              14137.2620              13281.2552   \n",
       "4             327164.9675             361262.1060             376591.8864   \n",
       "\n",
       "   Patient group:Polyp.75  \n",
       "0             510398.3294  \n",
       "1              29862.9664  \n",
       "2             216277.0586  \n",
       "3              12653.4161  \n",
       "4             368266.5694  \n",
       "\n",
       "[5 rows x 227 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(crc_table.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reset index position and name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "crc_table.set_index('Unnamed: 0', inplace=True)\n",
    "crc_table.index.name = 'row_id'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get the annotation dataframe using the index and the KEGG IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entity_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C02494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C05127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C00526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C00847</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       entity_id\n",
       "row_id          \n",
       "1         C02494\n",
       "2         C05127\n",
       "3         C00956\n",
       "4         C00526\n",
       "5         C00847"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "crc_annotation_df = crc_table[['Kegg Id']]\n",
    "\n",
    "crc_annotation_df.columns = ['entity_id']\n",
    "display(crc_annotation_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Drop columns not required by the intensity DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient group:CRC</th>\n",
       "      <th>Patient group:CRC.1</th>\n",
       "      <th>Patient group:CRC.2</th>\n",
       "      <th>Patient group:CRC.3</th>\n",
       "      <th>Patient group:CRC.4</th>\n",
       "      <th>Patient group:CRC.5</th>\n",
       "      <th>Patient group:CRC.6</th>\n",
       "      <th>Patient group:CRC.7</th>\n",
       "      <th>Patient group:CRC.8</th>\n",
       "      <th>Patient group:CRC.9</th>\n",
       "      <th>...</th>\n",
       "      <th>Patient group:Polyp.66</th>\n",
       "      <th>Patient group:Polyp.67</th>\n",
       "      <th>Patient group:Polyp.68</th>\n",
       "      <th>Patient group:Polyp.69</th>\n",
       "      <th>Patient group:Polyp.70</th>\n",
       "      <th>Patient group:Polyp.71</th>\n",
       "      <th>Patient group:Polyp.72</th>\n",
       "      <th>Patient group:Polyp.73</th>\n",
       "      <th>Patient group:Polyp.74</th>\n",
       "      <th>Patient group:Polyp.75</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>390953.9725</td>\n",
       "      <td>567984.4359</td>\n",
       "      <td>558842.7545</td>\n",
       "      <td>476949.0000</td>\n",
       "      <td>398317.0000</td>\n",
       "      <td>411224.0000</td>\n",
       "      <td>412250.6387</td>\n",
       "      <td>414501.1114</td>\n",
       "      <td>416166.0000</td>\n",
       "      <td>384549.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>390493.7447</td>\n",
       "      <td>375952.1076</td>\n",
       "      <td>399195.9585</td>\n",
       "      <td>1.018448e+06</td>\n",
       "      <td>503210.6702</td>\n",
       "      <td>311604.8089</td>\n",
       "      <td>514713.0726</td>\n",
       "      <td>407366.5548</td>\n",
       "      <td>463101.8944</td>\n",
       "      <td>510398.3294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34627.3365</td>\n",
       "      <td>52845.9284</td>\n",
       "      <td>31507.3427</td>\n",
       "      <td>29397.0000</td>\n",
       "      <td>38877.0000</td>\n",
       "      <td>25493.0000</td>\n",
       "      <td>44478.1854</td>\n",
       "      <td>27448.7350</td>\n",
       "      <td>34757.0000</td>\n",
       "      <td>26494.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>18266.0890</td>\n",
       "      <td>30752.9780</td>\n",
       "      <td>41086.0892</td>\n",
       "      <td>2.094303e+04</td>\n",
       "      <td>27630.5961</td>\n",
       "      <td>30531.1781</td>\n",
       "      <td>24293.0652</td>\n",
       "      <td>29915.2947</td>\n",
       "      <td>32606.5463</td>\n",
       "      <td>29862.9664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>141257.3640</td>\n",
       "      <td>528024.6865</td>\n",
       "      <td>188272.3016</td>\n",
       "      <td>107076.0918</td>\n",
       "      <td>158299.0710</td>\n",
       "      <td>136600.4976</td>\n",
       "      <td>235936.4772</td>\n",
       "      <td>419826.6337</td>\n",
       "      <td>351043.7015</td>\n",
       "      <td>124681.8762</td>\n",
       "      <td>...</td>\n",
       "      <td>149966.9484</td>\n",
       "      <td>188581.2257</td>\n",
       "      <td>153364.1518</td>\n",
       "      <td>2.666709e+06</td>\n",
       "      <td>228074.4199</td>\n",
       "      <td>132828.4060</td>\n",
       "      <td>215838.8101</td>\n",
       "      <td>179582.5152</td>\n",
       "      <td>227973.0218</td>\n",
       "      <td>216277.0586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13115.8131</td>\n",
       "      <td>14327.6960</td>\n",
       "      <td>12756.8649</td>\n",
       "      <td>10539.0000</td>\n",
       "      <td>11689.0000</td>\n",
       "      <td>12425.0000</td>\n",
       "      <td>15692.6312</td>\n",
       "      <td>15743.5808</td>\n",
       "      <td>8418.0000</td>\n",
       "      <td>9693.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>11790.1935</td>\n",
       "      <td>11041.2359</td>\n",
       "      <td>13050.2462</td>\n",
       "      <td>1.094457e+04</td>\n",
       "      <td>13652.0542</td>\n",
       "      <td>9250.8437</td>\n",
       "      <td>14597.4711</td>\n",
       "      <td>14137.2620</td>\n",
       "      <td>13281.2552</td>\n",
       "      <td>12653.4161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>333815.0605</td>\n",
       "      <td>402460.7838</td>\n",
       "      <td>389636.3191</td>\n",
       "      <td>373058.7101</td>\n",
       "      <td>323706.9435</td>\n",
       "      <td>358042.2763</td>\n",
       "      <td>355881.3356</td>\n",
       "      <td>347876.8878</td>\n",
       "      <td>347660.2188</td>\n",
       "      <td>354733.2473</td>\n",
       "      <td>...</td>\n",
       "      <td>379298.6241</td>\n",
       "      <td>356057.1511</td>\n",
       "      <td>412505.9735</td>\n",
       "      <td>4.260027e+05</td>\n",
       "      <td>358592.6192</td>\n",
       "      <td>314369.6151</td>\n",
       "      <td>327164.9675</td>\n",
       "      <td>361262.1060</td>\n",
       "      <td>376591.8864</td>\n",
       "      <td>368266.5694</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 224 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Patient group:CRC  Patient group:CRC.1  Patient group:CRC.2  \\\n",
       "row_id                                                                \n",
       "1             390953.9725          567984.4359          558842.7545   \n",
       "2              34627.3365           52845.9284           31507.3427   \n",
       "3             141257.3640          528024.6865          188272.3016   \n",
       "4              13115.8131           14327.6960           12756.8649   \n",
       "5             333815.0605          402460.7838          389636.3191   \n",
       "\n",
       "        Patient group:CRC.3  Patient group:CRC.4  Patient group:CRC.5  \\\n",
       "row_id                                                                  \n",
       "1               476949.0000          398317.0000          411224.0000   \n",
       "2                29397.0000           38877.0000           25493.0000   \n",
       "3               107076.0918          158299.0710          136600.4976   \n",
       "4                10539.0000           11689.0000           12425.0000   \n",
       "5               373058.7101          323706.9435          358042.2763   \n",
       "\n",
       "        Patient group:CRC.6  Patient group:CRC.7  Patient group:CRC.8  \\\n",
       "row_id                                                                  \n",
       "1               412250.6387          414501.1114          416166.0000   \n",
       "2                44478.1854           27448.7350           34757.0000   \n",
       "3               235936.4772          419826.6337          351043.7015   \n",
       "4                15692.6312           15743.5808            8418.0000   \n",
       "5               355881.3356          347876.8878          347660.2188   \n",
       "\n",
       "        Patient group:CRC.9  ...  Patient group:Polyp.66  \\\n",
       "row_id                       ...                           \n",
       "1               384549.0000  ...             390493.7447   \n",
       "2                26494.0000  ...              18266.0890   \n",
       "3               124681.8762  ...             149966.9484   \n",
       "4                 9693.0000  ...              11790.1935   \n",
       "5               354733.2473  ...             379298.6241   \n",
       "\n",
       "        Patient group:Polyp.67  Patient group:Polyp.68  \\\n",
       "row_id                                                   \n",
       "1                  375952.1076             399195.9585   \n",
       "2                   30752.9780              41086.0892   \n",
       "3                  188581.2257             153364.1518   \n",
       "4                   11041.2359              13050.2462   \n",
       "5                  356057.1511             412505.9735   \n",
       "\n",
       "        Patient group:Polyp.69  Patient group:Polyp.70  \\\n",
       "row_id                                                   \n",
       "1                 1.018448e+06             503210.6702   \n",
       "2                 2.094303e+04              27630.5961   \n",
       "3                 2.666709e+06             228074.4199   \n",
       "4                 1.094457e+04              13652.0542   \n",
       "5                 4.260027e+05             358592.6192   \n",
       "\n",
       "        Patient group:Polyp.71  Patient group:Polyp.72  \\\n",
       "row_id                                                   \n",
       "1                  311604.8089             514713.0726   \n",
       "2                   30531.1781              24293.0652   \n",
       "3                  132828.4060             215838.8101   \n",
       "4                    9250.8437              14597.4711   \n",
       "5                  314369.6151             327164.9675   \n",
       "\n",
       "        Patient group:Polyp.73  Patient group:Polyp.74  Patient group:Polyp.75  \n",
       "row_id                                                                          \n",
       "1                  407366.5548             463101.8944             510398.3294  \n",
       "2                   29915.2947              32606.5463              29862.9664  \n",
       "3                  179582.5152             227973.0218             216277.0586  \n",
       "4                   14137.2620              13281.2552              12653.4161  \n",
       "5                  361262.1060             376591.8864             368266.5694  \n",
       "\n",
       "[5 rows x 224 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "crc_int_df = crc_table.drop(columns=['Metabolites', 'Kegg Id'])\n",
    "display(crc_int_df.head())\n",
    "columns = crc_int_df.columns\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make group lists for the experimental design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "crc=[]\n",
    "healthy = []\n",
    "polyp=[]\n",
    "for c in columns:\n",
    "    if 'CRC' in c:\n",
    "        crc.append(c)\n",
    "    elif 'Healthy' in c:\n",
    "        healthy.append(c)\n",
    "    elif 'Polyp' in c:\n",
    "        polyp.append(c)\n",
    "crc_experimental_design = {'comparisons': [{'case': 'CRC', 'control': 'healthy', 'name': 'crc/healthy'},\n",
    "  {'case': 'polyp', 'control': 'healthy', 'name': 'crc/healthy'}]}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "crc_experimental_design['groups']={'healthy':healthy, 'CRC':crc, 'polyp': polyp}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-20 10:59:09.429 | DEBUG    | pals.feature_extraction:__init__:40 - Loading /Users/Karen/PALS/pals/data/PiMP_KEGG.json.zip\n",
      "2019-12-20 10:59:09.478 | DEBUG    | pals.feature_extraction:__init__:92 - Mapping pathway to unique ids\n",
      "2019-12-20 10:59:09.486 | DEBUG    | pals.feature_extraction:__init__:106 - Creating dataset to pathway mapping\n",
      "2019-12-20 10:59:09.498 | DEBUG    | pals.feature_extraction:__init__:134 - Computing unique id counts\n"
     ]
    }
   ],
   "source": [
    "crc_ds = DataSource(crc_int_df, crc_annotation_df, crc_experimental_design, DATABASE_PIMP_KEGG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-20 10:59:12.158 | DEBUG    | pals.pathway_analysis:_change_zero_peak_ints:385 - Setting the zero intensity values in the dataframe\n",
      "2019-12-20 10:59:12.263 | DEBUG    | pals.pathway_analysis:_standardize_intensity_df:366 - Scaling the data across the sample: zero mean and unit variance\n",
      "2019-12-20 10:59:12.267 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:205 - Mean values of the rows in the DF is [-0. -0. -0.  0. -0. -0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.  0. -0.\n",
      " -0.  0.  0. -0. -0.  0. -0.  0. -0. -0. -0.  0. -0.  0.  0.  0. -0. -0.\n",
      "  0.  0.  0. -0. -0. -0. -0.  0.  0. -0. -0. -0.  0. -0.  0. -0.  0.  0.\n",
      "  0. -0.  0. -0.  0.  0. -0. -0. -0. -0.  0. -0. -0.  0.  0.  0. -0.  0.\n",
      " -0. -0. -0.  0.  0.  0. -0.  0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.\n",
      "  0.  0. -0.  0. -0. -0. -0. -0. -0. -0.  0. -0.  0. -0. -0.]\n",
      "2019-12-20 10:59:12.268 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:206 - Variance in the rows of the DF is [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.\n",
      " 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.\n",
      " 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.\n",
      " 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.\n",
      " 1. 1. 1. 1. 1. 1. 1. 1. 1.]\n",
      "2019-12-20 10:59:18.991 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:217 - Calculating plage p-values with resampling\n",
      "2019-12-20 10:59:18.992 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison crc/healthy\n",
      "2019-12-20 10:59:18.993 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-20 10:59:19.177 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-20 10:59:19.326 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-20 10:59:19.478 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-20 10:59:19.634 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-20 10:59:19.802 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-20 10:59:19.963 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-20 10:59:20.122 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-20 10:59:20.277 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-20 10:59:20.434 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-20 10:59:20.594 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 1\n",
      "2019-12-20 10:59:20.729 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison crc/healthy\n",
      "2019-12-20 10:59:20.730 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-20 10:59:20.885 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-20 10:59:21.048 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-20 10:59:21.224 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-20 10:59:21.400 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-20 10:59:21.565 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-20 10:59:21.719 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-20 10:59:21.872 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-20 10:59:22.028 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-20 10:59:22.188 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-20 10:59:22.353 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 1\n",
      "2019-12-20 10:59:22.510 | DEBUG    | pals.pathway_analysis:_calculate_coverage_df:471 - Calculating dataset formula coverage\n",
      "2019-12-20 10:59:22.515 | DEBUG    | pals.pathway_analysis:calculate_hg_values:296 - Calculating the hyper-geometric p-values\n",
      "2019-12-20 10:59:22.705 | DEBUG    | pals.pathway_analysis:calculate_hg_values:328 - Calculating the combined p-values\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-32-15227565db28>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mpals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPALS\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcrc_ds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_replace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplage_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhg_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mpathway_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpals\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_pathway_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/PALS/pals/pathway_analysis.py\u001b[0m in \u001b[0;36mget_pathway_df\u001b[0;34m(self, resample, standardize)\u001b[0m\n\u001b[1;32m    180\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    181\u001b[0m             \u001b[0mplage_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_up_plage_p_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mactivity_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 182\u001b[0;31m         \u001b[0mpathway_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcalculate_hg_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mplage_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    183\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mpathway_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/PALS/pals/pathway_analysis.py\u001b[0m in \u001b[0;36mcalculate_hg_values\u001b[0;34m(self, pathway_df)\u001b[0m\n\u001b[1;32m    338\u001b[0m                 \u001b[0mp_value\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpathway_df_merge\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmp\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mp_value_colname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    339\u001b[0m                 \u001b[0msf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpathway_df_merge\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmp\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sf'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 340\u001b[0;31m                 \u001b[0mcom_p\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcombine_pvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mp_value\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msf\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'stouffer'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplage_weight\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhg_weight\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    341\u001b[0m                 \u001b[0mcombine_p_pathway\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcom_p\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    342\u001b[0m             \u001b[0mcombine_p_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcombine_p_pathway\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/share/virtualenvs/PALS-d7uEJZs8/lib/python3.7/site-packages/scipy/stats/stats.py\u001b[0m in \u001b[0;36mcombine_pvalues\u001b[0;34m(pvalues, method, weights)\u001b[0m\n\u001b[1;32m   6192\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"weights is not 1-D\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6193\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6194\u001b[0;31m         \u001b[0mZi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdistributions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnorm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6195\u001b[0m         \u001b[0mstatistic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mweights\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mZi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlinalg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnorm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mweights\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6196\u001b[0m         \u001b[0mpval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdistributions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnorm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstatistic\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/share/virtualenvs/PALS-d7uEJZs8/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py\u001b[0m in \u001b[0;36misf\u001b[0;34m(self, q, *args, **kwds)\u001b[0m\n\u001b[1;32m   2021\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2022\u001b[0m         \u001b[0mcond0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_argcheck\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mscale\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2023\u001b[0;31m         \u001b[0mcond1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2024\u001b[0m         \u001b[0mcond2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcond0\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2025\u001b[0m         \u001b[0mcond3\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcond0\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/share/virtualenvs/PALS-d7uEJZs8/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__nonzero__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1553\u001b[0m             \u001b[0;34m\"The truth value of a {0} is ambiguous. \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1554\u001b[0m             \"Use a.empty, a.bool(), a.item(), a.any() or a.all().\".format(\n\u001b[0;32m-> 1555\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1556\u001b[0m             )\n\u001b[1;32m   1557\u001b[0m         )\n",
      "\u001b[0;31mValueError\u001b[0m: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all()."
     ]
    }
   ],
   "source": [
    "pals = PALS(crc_ds, min_replace=5000, plage_weight=5, hg_weight=1)\n",
    "pathway_df = pals.get_pathway_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Back to Beer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'comparisons': [{'case': 'beer1', 'control': 'beer2', 'name': 'beer1/beer2'},\n",
       "  {'case': 'beer3', 'control': 'beer4', 'name': 'beer3/beer4'}],\n",
       " 'groups': {'beer4': ['Beer_4_full3.mzXML',\n",
       "   'Beer_4_full2.mzXML',\n",
       "   'Beer_4_full1.mzXML'],\n",
       "  'beer3': ['Beer_3_full3.mzXML', 'Beer_3_full2.mzXML', 'Beer_3_full1.mzXML'],\n",
       "  'beer2': ['Beer_2_full3.mzXML', 'Beer_2_full1.mzXML', 'Beer_2_full2.mzXML'],\n",
       "  'beer1': ['Beer_1_full2.mzXML', 'Beer_1_full1.mzXML', 'Beer_1_full3.mzXML']}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experimental_design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PALS analysis using KEGG database exported from PiMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entity_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3033929</th>\n",
       "      <td>C00148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3036581</th>\n",
       "      <td>C00148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3036855</th>\n",
       "      <td>C00148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3038249</th>\n",
       "      <td>C00148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3033929</th>\n",
       "      <td>C00163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040926</th>\n",
       "      <td>C20522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3040929</th>\n",
       "      <td>C20582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3041077</th>\n",
       "      <td>C20499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3041172</th>\n",
       "      <td>C20504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3041172</th>\n",
       "      <td>C20502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14549 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        entity_id\n",
       "row_id           \n",
       "3033929    C00148\n",
       "3036581    C00148\n",
       "3036855    C00148\n",
       "3038249    C00148\n",
       "3033929    C00163\n",
       "...           ...\n",
       "3040926    C20522\n",
       "3040929    C20582\n",
       "3041077    C20499\n",
       "3041172    C20504\n",
       "3041172    C20502\n",
       "\n",
       "[14549 rows x 1 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:29:54.631 | DEBUG    | pals.feature_extraction:get_data:96 - Loading C:\\Users\\joewa\\Work\\git\\PALS\\pals\\data\\PiMP_KEGG.json.zip\n",
      "2019-12-18 13:29:54.657 | DEBUG    | pals.feature_extraction:__init__:46 - Mapping pathway to unique ids\n",
      "2019-12-18 13:29:54.663 | DEBUG    | pals.feature_extraction:__init__:60 - Creating dataset to pathway mapping\n",
      "2019-12-18 13:29:55.564 | DEBUG    | pals.feature_extraction:__init__:88 - Computing unique id counts\n"
     ]
    }
   ],
   "source": [
    "ds = DataSource(int_df, annotation_df, experimental_design, DATABASE_PIMP_KEGG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:29:55.640 | DEBUG    | pals.pathway_analysis:_change_zero_peak_ints:385 - Setting the zero intensity values in the dataframe\n",
      "2019-12-18 13:29:55.670 | DEBUG    | pals.pathway_analysis:_standardize_intensity_df:366 - Scaling the data across the sample: zero mean and unit variance\n",
      "2019-12-18 13:29:55.677 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:205 - Mean values of the rows in the DF is [ 0.  0. -0. ... -0. -0. -0.]\n",
      "2019-12-18 13:29:55.678 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:206 - Variance in the rows of the DF is [1. 1. 1. ... 1. 1. 1.]\n",
      "2019-12-18 13:29:56.386 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:217 - Calculating plage p-values with resampling\n",
      "2019-12-18 13:29:56.387 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison beer1/beer2\n",
      "2019-12-18 13:29:56.387 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-18 13:29:56.486 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-18 13:29:56.578 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-18 13:29:56.674 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-18 13:29:56.769 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-18 13:29:56.863 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-18 13:29:56.974 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-18 13:29:57.070 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-18 13:29:57.170 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-18 13:29:57.265 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-18 13:29:57.365 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 0\n",
      "2019-12-18 13:29:57.536 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison beer3/beer4\n",
      "2019-12-18 13:29:57.537 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-18 13:29:57.638 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-18 13:29:57.734 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-18 13:29:57.834 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-18 13:29:57.928 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-18 13:29:58.024 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-18 13:29:58.120 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-18 13:29:58.214 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-18 13:29:58.318 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-18 13:29:58.416 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-18 13:29:58.518 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 0\n",
      "c:\\users\\joewa\\.virtualenvs\\pals-4ls44jcw\\lib\\site-packages\\scipy\\stats\\_distn_infrastructure.py:901: RuntimeWarning: invalid value encountered in less\n",
      "  return (a < x) & (x < b)\n",
      "c:\\users\\joewa\\.virtualenvs\\pals-4ls44jcw\\lib\\site-packages\\scipy\\stats\\_distn_infrastructure.py:1892: RuntimeWarning: invalid value encountered in less_equal\n",
      "  cond2 = cond0 & (x <= _a)\n",
      "2019-12-18 13:29:58.674 | DEBUG    | pals.pathway_analysis:_calculate_coverage_df:471 - Calculating dataset formula coverage\n",
      "2019-12-18 13:29:58.677 | DEBUG    | pals.pathway_analysis:calculate_hg_values:296 - Calculating the hyper-geometric p-values\n",
      "2019-12-18 13:29:58.834 | DEBUG    | pals.pathway_analysis:calculate_hg_values:328 - Calculating the combined p-values\n"
     ]
    }
   ],
   "source": [
    "pals = PALS(ds, min_replace=5000, plage_weight=5, hg_weight=1)\n",
    "pathway_df = pals.get_pathway_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw_name</th>\n",
       "      <th>beer1/beer2 p-value</th>\n",
       "      <th>beer3/beer4 p-value</th>\n",
       "      <th>unq_pw_F</th>\n",
       "      <th>tot_ds_F</th>\n",
       "      <th>F_coverage</th>\n",
       "      <th>sf</th>\n",
       "      <th>exp_F</th>\n",
       "      <th>Ex_Cov</th>\n",
       "      <th>PiMP_KEGG beer1/beer2 comb_p</th>\n",
       "      <th>PiMP_KEGG beer3/beer4 comb_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>map00380</th>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "      <td>0.026410</td>\n",
       "      <td>0.050835</td>\n",
       "      <td>64</td>\n",
       "      <td>34</td>\n",
       "      <td>53.12</td>\n",
       "      <td>2.024112e-08</td>\n",
       "      <td>14.26</td>\n",
       "      <td>22.28</td>\n",
       "      <td>0.001464</td>\n",
       "      <td>0.003665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00330</th>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>0.076544</td>\n",
       "      <td>0.031235</td>\n",
       "      <td>79</td>\n",
       "      <td>50</td>\n",
       "      <td>63.29</td>\n",
       "      <td>7.494293e-16</td>\n",
       "      <td>17.61</td>\n",
       "      <td>22.29</td>\n",
       "      <td>0.001512</td>\n",
       "      <td>0.000348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00460</th>\n",
       "      <td>Cyanoamino acid metabolism</td>\n",
       "      <td>0.033922</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>40</td>\n",
       "      <td>25</td>\n",
       "      <td>62.50</td>\n",
       "      <td>1.448042e-08</td>\n",
       "      <td>8.92</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.001998</td>\n",
       "      <td>0.002702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00300</th>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>0.037750</td>\n",
       "      <td>0.034624</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>74.07</td>\n",
       "      <td>3.670262e-09</td>\n",
       "      <td>6.02</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.002007</td>\n",
       "      <td>0.001775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00400</th>\n",
       "      <td>Phenylalanine, tyrosine and tryptophan biosynt...</td>\n",
       "      <td>0.051530</td>\n",
       "      <td>0.031751</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>73.33</td>\n",
       "      <td>9.098577e-10</td>\n",
       "      <td>6.69</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.002736</td>\n",
       "      <td>0.001355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map07227</th>\n",
       "      <td>Histamine H2/H3 receptor agonists/antagonists</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.212082</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10.00</td>\n",
       "      <td>7.411824e-01</td>\n",
       "      <td>2.23</td>\n",
       "      <td>22.30</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.255651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00523</th>\n",
       "      <td>Polyketide sugar unit biosynthesis</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.053513</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>10.53</td>\n",
       "      <td>8.566036e-01</td>\n",
       "      <td>4.23</td>\n",
       "      <td>22.26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.085109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map05032</th>\n",
       "      <td>Morphine addiction</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.071093</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>37.50</td>\n",
       "      <td>1.185365e-01</td>\n",
       "      <td>1.78</td>\n",
       "      <td>22.25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.047353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map01040</th>\n",
       "      <td>Biosynthesis of unsaturated fatty acids</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.069916</td>\n",
       "      <td>42</td>\n",
       "      <td>4</td>\n",
       "      <td>9.52</td>\n",
       "      <td>9.770476e-01</td>\n",
       "      <td>9.36</td>\n",
       "      <td>22.29</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.145428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00351</th>\n",
       "      <td>DDT degradation</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.233911</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>10.00</td>\n",
       "      <td>8.790777e-01</td>\n",
       "      <td>4.46</td>\n",
       "      <td>22.30</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.314762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>226 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    pw_name  \\\n",
       "map00380                              Tryptophan metabolism   \n",
       "map00330                    Arginine and proline metabolism   \n",
       "map00460                         Cyanoamino acid metabolism   \n",
       "map00300                                Lysine biosynthesis   \n",
       "map00400  Phenylalanine, tyrosine and tryptophan biosynt...   \n",
       "...                                                     ...   \n",
       "map07227      Histamine H2/H3 receptor agonists/antagonists   \n",
       "map00523                 Polyketide sugar unit biosynthesis   \n",
       "map05032                                 Morphine addiction   \n",
       "map01040            Biosynthesis of unsaturated fatty acids   \n",
       "map00351                                    DDT degradation   \n",
       "\n",
       "          beer1/beer2 p-value  beer3/beer4 p-value  unq_pw_F  tot_ds_F  \\\n",
       "map00380             0.026410             0.050835        64        34   \n",
       "map00330             0.076544             0.031235        79        50   \n",
       "map00460             0.033922             0.042044        40        25   \n",
       "map00300             0.037750             0.034624        27        20   \n",
       "map00400             0.051530             0.031751        30        22   \n",
       "...                       ...                  ...       ...       ...   \n",
       "map07227             1.000000             0.212082        10         1   \n",
       "map00523             1.000000             0.053513        19         2   \n",
       "map05032             1.000000             0.071093         8         3   \n",
       "map01040             1.000000             0.069916        42         4   \n",
       "map00351             1.000000             0.233911        20         2   \n",
       "\n",
       "          F_coverage            sf  exp_F  Ex_Cov  \\\n",
       "map00380       53.12  2.024112e-08  14.26   22.28   \n",
       "map00330       63.29  7.494293e-16  17.61   22.29   \n",
       "map00460       62.50  1.448042e-08   8.92   22.30   \n",
       "map00300       74.07  3.670262e-09   6.02   22.30   \n",
       "map00400       73.33  9.098577e-10   6.69   22.30   \n",
       "...              ...           ...    ...     ...   \n",
       "map07227       10.00  7.411824e-01   2.23   22.30   \n",
       "map00523       10.53  8.566036e-01   4.23   22.26   \n",
       "map05032       37.50  1.185365e-01   1.78   22.25   \n",
       "map01040        9.52  9.770476e-01   9.36   22.29   \n",
       "map00351       10.00  8.790777e-01   4.46   22.30   \n",
       "\n",
       "          PiMP_KEGG beer1/beer2 comb_p  PiMP_KEGG beer3/beer4 comb_p  \n",
       "map00380                      0.001464                      0.003665  \n",
       "map00330                      0.001512                      0.000348  \n",
       "map00460                      0.001998                      0.002702  \n",
       "map00300                      0.002007                      0.001775  \n",
       "map00400                      0.002736                      0.001355  \n",
       "...                                ...                           ...  \n",
       "map07227                      1.000000                      0.255651  \n",
       "map00523                      1.000000                      0.085109  \n",
       "map05032                      1.000000                      0.047353  \n",
       "map01040                      1.000000                      0.145428  \n",
       "map00351                      1.000000                      0.314762  \n",
       "\n",
       "[226 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df.sort_values('PiMP_KEGG beer1/beer2 comb_p', ascending=True, inplace=True)\n",
    "pathway_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PALS analysis using KEGG database exported from Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:29:59.211 | DEBUG    | pals.feature_extraction:get_data:120 - Loading ..\\pals\\data\\reactome\\metabolic_pathways\\COMPOUND\\Homo sapiens.json.zip\n",
      "2019-12-18 13:29:59.234 | DEBUG    | pals.feature_extraction:__init__:46 - Mapping pathway to unique ids\n",
      "2019-12-18 13:29:59.236 | DEBUG    | pals.feature_extraction:__init__:60 - Creating dataset to pathway mapping\n",
      "2019-12-18 13:30:00.161 | DEBUG    | pals.feature_extraction:__init__:88 - Computing unique id counts\n"
     ]
    }
   ],
   "source": [
    "ds = DataSource(int_df, annotation_df, experimental_design, DATABASE_REACTOME_KEGG, \n",
    "                reactome_species=REACTOME_SPECIES_HOMO_SAPIENS, reactome_metabolic_pathway_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:30:00.247 | DEBUG    | pals.pathway_analysis:_change_zero_peak_ints:385 - Setting the zero intensity values in the dataframe\n",
      "2019-12-18 13:30:00.276 | DEBUG    | pals.pathway_analysis:_standardize_intensity_df:366 - Scaling the data across the sample: zero mean and unit variance\n",
      "2019-12-18 13:30:00.285 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:205 - Mean values of the rows in the DF is [ 0.  0. -0. ... -0. -0. -0.]\n",
      "2019-12-18 13:30:00.286 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:206 - Variance in the rows of the DF is [1. 1. 1. ... 1. 1. 1.]\n",
      "2019-12-18 13:30:00.725 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:217 - Calculating plage p-values with resampling\n",
      "2019-12-18 13:30:00.726 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison beer1/beer2\n",
      "2019-12-18 13:30:00.727 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-18 13:30:00.852 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-18 13:30:00.979 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-18 13:30:01.113 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-18 13:30:01.255 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-18 13:30:01.411 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-18 13:30:01.591 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-18 13:30:01.704 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-18 13:30:01.801 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-18 13:30:01.894 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-18 13:30:01.987 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 1\n",
      "2019-12-18 13:30:02.149 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison beer3/beer4\n",
      "2019-12-18 13:30:02.150 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-18 13:30:02.254 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-18 13:30:02.349 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-18 13:30:02.456 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-18 13:30:02.558 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-18 13:30:02.665 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-18 13:30:02.759 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-18 13:30:02.869 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-18 13:30:02.962 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-18 13:30:03.062 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-18 13:30:03.163 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 1\n",
      "2019-12-18 13:30:03.299 | DEBUG    | pals.pathway_analysis:_calculate_coverage_df:471 - Calculating dataset formula coverage\n",
      "2019-12-18 13:30:03.303 | DEBUG    | pals.pathway_analysis:calculate_hg_values:296 - Calculating the hyper-geometric p-values\n",
      "2019-12-18 13:30:03.424 | DEBUG    | pals.pathway_analysis:calculate_hg_values:328 - Calculating the combined p-values\n"
     ]
    }
   ],
   "source": [
    "pals = PALS(ds, min_replace=5000, plage_weight=5, hg_weight=1)\n",
    "pathway_df = pals.get_pathway_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw_name</th>\n",
       "      <th>beer1/beer2 p-value</th>\n",
       "      <th>beer3/beer4 p-value</th>\n",
       "      <th>unq_pw_F</th>\n",
       "      <th>tot_ds_F</th>\n",
       "      <th>F_coverage</th>\n",
       "      <th>sf</th>\n",
       "      <th>exp_F</th>\n",
       "      <th>Ex_Cov</th>\n",
       "      <th>COMPOUND beer1/beer2 comb_p</th>\n",
       "      <th>COMPOUND beer3/beer4 comb_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R-HSA-71240</th>\n",
       "      <td>Tryptophan catabolism</td>\n",
       "      <td>0.016712</td>\n",
       "      <td>0.057485</td>\n",
       "      <td>27</td>\n",
       "      <td>14</td>\n",
       "      <td>51.85</td>\n",
       "      <td>0.020106</td>\n",
       "      <td>9.12</td>\n",
       "      <td>33.78</td>\n",
       "      <td>0.006423</td>\n",
       "      <td>0.025709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-2024096</th>\n",
       "      <td>HS-GAG degradation</td>\n",
       "      <td>0.011545</td>\n",
       "      <td>0.078462</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>20.00</td>\n",
       "      <td>0.659417</td>\n",
       "      <td>1.69</td>\n",
       "      <td>33.80</td>\n",
       "      <td>0.015887</td>\n",
       "      <td>0.095534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-163685</th>\n",
       "      <td>Integration of energy metabolism</td>\n",
       "      <td>0.032313</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.113509</td>\n",
       "      <td>0.34</td>\n",
       "      <td>34.00</td>\n",
       "      <td>0.020237</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-351143</th>\n",
       "      <td>Agmatine biosynthesis</td>\n",
       "      <td>0.023579</td>\n",
       "      <td>0.115855</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>40.00</td>\n",
       "      <td>0.327631</td>\n",
       "      <td>1.69</td>\n",
       "      <td>33.80</td>\n",
       "      <td>0.020981</td>\n",
       "      <td>0.103781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-71182</th>\n",
       "      <td>Phenylalanine and tyrosine catabolism</td>\n",
       "      <td>0.051213</td>\n",
       "      <td>0.131596</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>54.17</td>\n",
       "      <td>0.015294</td>\n",
       "      <td>8.11</td>\n",
       "      <td>33.79</td>\n",
       "      <td>0.021405</td>\n",
       "      <td>0.064101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-1989781</th>\n",
       "      <td>PPARA activates gene expression</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.341515</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>50.00</td>\n",
       "      <td>0.264649</td>\n",
       "      <td>0.68</td>\n",
       "      <td>34.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.300218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-2046106</th>\n",
       "      <td>alpha-linolenic acid (ALA) metabolism</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.094125</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>7.14</td>\n",
       "      <td>0.983784</td>\n",
       "      <td>4.73</td>\n",
       "      <td>33.79</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.191954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-1855183</th>\n",
       "      <td>Synthesis of IP2, IP, and Ins in the cytosol</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.064520</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>27.27</td>\n",
       "      <td>0.622386</td>\n",
       "      <td>3.72</td>\n",
       "      <td>33.82</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.076749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-70221</th>\n",
       "      <td>Glycogen breakdown (glycogenolysis)</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.064520</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>30.00</td>\n",
       "      <td>0.541649</td>\n",
       "      <td>3.38</td>\n",
       "      <td>33.80</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.071064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-71384</th>\n",
       "      <td>Ethanol oxidation</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.094125</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>7.69</td>\n",
       "      <td>0.976586</td>\n",
       "      <td>4.39</td>\n",
       "      <td>33.77</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.183962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>167 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    pw_name  \\\n",
       "R-HSA-71240                           Tryptophan catabolism   \n",
       "R-HSA-2024096                            HS-GAG degradation   \n",
       "R-HSA-163685               Integration of energy metabolism   \n",
       "R-HSA-351143                          Agmatine biosynthesis   \n",
       "R-HSA-71182           Phenylalanine and tyrosine catabolism   \n",
       "...                                                     ...   \n",
       "R-HSA-1989781               PPARA activates gene expression   \n",
       "R-HSA-2046106         alpha-linolenic acid (ALA) metabolism   \n",
       "R-HSA-1855183  Synthesis of IP2, IP, and Ins in the cytosol   \n",
       "R-HSA-70221             Glycogen breakdown (glycogenolysis)   \n",
       "R-HSA-71384                               Ethanol oxidation   \n",
       "\n",
       "               beer1/beer2 p-value  beer3/beer4 p-value  unq_pw_F  tot_ds_F  \\\n",
       "R-HSA-71240               0.016712             0.057485        27        14   \n",
       "R-HSA-2024096             0.011545             0.078462         5         1   \n",
       "R-HSA-163685              0.032313             1.000000         1         1   \n",
       "R-HSA-351143              0.023579             0.115855         5         2   \n",
       "R-HSA-71182               0.051213             0.131596        24        13   \n",
       "...                            ...                  ...       ...       ...   \n",
       "R-HSA-1989781             1.000000             0.341515         2         1   \n",
       "R-HSA-2046106             1.000000             0.094125        14         1   \n",
       "R-HSA-1855183             1.000000             0.064520        11         3   \n",
       "R-HSA-70221               1.000000             0.064520        10         3   \n",
       "R-HSA-71384               1.000000             0.094125        13         1   \n",
       "\n",
       "               F_coverage        sf  exp_F  Ex_Cov  \\\n",
       "R-HSA-71240         51.85  0.020106   9.12   33.78   \n",
       "R-HSA-2024096       20.00  0.659417   1.69   33.80   \n",
       "R-HSA-163685       100.00  0.113509   0.34   34.00   \n",
       "R-HSA-351143        40.00  0.327631   1.69   33.80   \n",
       "R-HSA-71182         54.17  0.015294   8.11   33.79   \n",
       "...                   ...       ...    ...     ...   \n",
       "R-HSA-1989781       50.00  0.264649   0.68   34.00   \n",
       "R-HSA-2046106        7.14  0.983784   4.73   33.79   \n",
       "R-HSA-1855183       27.27  0.622386   3.72   33.82   \n",
       "R-HSA-70221         30.00  0.541649   3.38   33.80   \n",
       "R-HSA-71384          7.69  0.976586   4.39   33.77   \n",
       "\n",
       "               COMPOUND beer1/beer2 comb_p  COMPOUND beer3/beer4 comb_p  \n",
       "R-HSA-71240                       0.006423                     0.025709  \n",
       "R-HSA-2024096                     0.015887                     0.095534  \n",
       "R-HSA-163685                      0.020237                     1.000000  \n",
       "R-HSA-351143                      0.020981                     0.103781  \n",
       "R-HSA-71182                       0.021405                     0.064101  \n",
       "...                                    ...                          ...  \n",
       "R-HSA-1989781                     1.000000                     0.300218  \n",
       "R-HSA-2046106                     1.000000                     0.191954  \n",
       "R-HSA-1855183                     1.000000                     0.076749  \n",
       "R-HSA-70221                       1.000000                     0.071064  \n",
       "R-HSA-71384                       1.000000                     0.183962  \n",
       "\n",
       "[167 rows x 11 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df.sort_values('COMPOUND beer1/beer2 comb_p', ascending=True, inplace=True)\n",
    "pathway_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PALS analysis of compounds by connecting to Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:30:03.748 | DEBUG    | pals.feature_extraction:get_data:102 - Retrieving data for Homo sapiens from Reactome COMPOUND metabolic_pathway_only=True\n",
      "2019-12-18 13:30:08.208 | DEBUG    | pals.feature_extraction:__init__:46 - Mapping pathway to unique ids\n",
      "2019-12-18 13:30:08.210 | DEBUG    | pals.feature_extraction:__init__:60 - Creating dataset to pathway mapping\n",
      "2019-12-18 13:30:09.122 | DEBUG    | pals.feature_extraction:__init__:88 - Computing unique id counts\n"
     ]
    }
   ],
   "source": [
    "ds = DataSource(int_df, annotation_df, experimental_design, DATABASE_REACTOME_KEGG, \n",
    "                reactome_species=REACTOME_SPECIES_HOMO_SAPIENS, reactome_metabolic_pathway_only=True, reactome_query=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:30:09.200 | DEBUG    | pals.pathway_analysis:_change_zero_peak_ints:385 - Setting the zero intensity values in the dataframe\n",
      "2019-12-18 13:30:09.229 | DEBUG    | pals.pathway_analysis:_standardize_intensity_df:366 - Scaling the data across the sample: zero mean and unit variance\n",
      "2019-12-18 13:30:09.238 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:205 - Mean values of the rows in the DF is [ 0.  0. -0. ... -0. -0. -0.]\n",
      "2019-12-18 13:30:09.239 | DEBUG    | pals.pathway_analysis:get_plage_activity_df:206 - Variance in the rows of the DF is [1. 1. 1. ... 1. 1. 1.]\n",
      "2019-12-18 13:30:09.645 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:217 - Calculating plage p-values with resampling\n",
      "2019-12-18 13:30:09.646 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison beer1/beer2\n",
      "2019-12-18 13:30:09.647 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-18 13:30:09.741 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-18 13:30:09.834 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-18 13:30:09.931 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-18 13:30:10.024 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-18 13:30:10.125 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-18 13:30:10.224 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-18 13:30:10.329 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-18 13:30:10.457 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-18 13:30:10.576 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-18 13:30:10.681 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 1\n",
      "2019-12-18 13:30:10.799 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:221 - Comparison beer3/beer4\n",
      "2019-12-18 13:30:10.800 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 0/1000\n",
      "2019-12-18 13:30:10.894 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 100/1000\n",
      "2019-12-18 13:30:10.986 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 200/1000\n",
      "2019-12-18 13:30:11.082 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 300/1000\n",
      "2019-12-18 13:30:11.176 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 400/1000\n",
      "2019-12-18 13:30:11.277 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 500/1000\n",
      "2019-12-18 13:30:11.369 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 600/1000\n",
      "2019-12-18 13:30:11.476 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 700/1000\n",
      "2019-12-18 13:30:11.570 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 800/1000\n",
      "2019-12-18 13:30:11.667 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:229 - Resampling 900/1000\n",
      "2019-12-18 13:30:11.758 | DEBUG    | pals.pathway_analysis:set_up_resample_plage_p_df:235 - Total time 0\n",
      "2019-12-18 13:30:11.916 | DEBUG    | pals.pathway_analysis:_calculate_coverage_df:471 - Calculating dataset formula coverage\n",
      "2019-12-18 13:30:11.919 | DEBUG    | pals.pathway_analysis:calculate_hg_values:296 - Calculating the hyper-geometric p-values\n",
      "2019-12-18 13:30:12.029 | DEBUG    | pals.pathway_analysis:calculate_hg_values:328 - Calculating the combined p-values\n"
     ]
    }
   ],
   "source": [
    "pals = PALS(ds, min_replace=5000, plage_weight=5, hg_weight=1)\n",
    "pathway_df = pals.get_pathway_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw_name</th>\n",
       "      <th>beer1/beer2 p-value</th>\n",
       "      <th>beer3/beer4 p-value</th>\n",
       "      <th>unq_pw_F</th>\n",
       "      <th>tot_ds_F</th>\n",
       "      <th>F_coverage</th>\n",
       "      <th>sf</th>\n",
       "      <th>exp_F</th>\n",
       "      <th>Ex_Cov</th>\n",
       "      <th>COMPOUND beer1/beer2 comb_p</th>\n",
       "      <th>COMPOUND beer3/beer4 comb_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R-HSA-71240</th>\n",
       "      <td>Tryptophan catabolism</td>\n",
       "      <td>0.025106</td>\n",
       "      <td>0.053509</td>\n",
       "      <td>27</td>\n",
       "      <td>14</td>\n",
       "      <td>51.85</td>\n",
       "      <td>0.020106</td>\n",
       "      <td>9.12</td>\n",
       "      <td>33.78</td>\n",
       "      <td>0.010104</td>\n",
       "      <td>0.023696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-2024096</th>\n",
       "      <td>HS-GAG degradation</td>\n",
       "      <td>0.018223</td>\n",
       "      <td>0.066147</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>20.00</td>\n",
       "      <td>0.659417</td>\n",
       "      <td>1.69</td>\n",
       "      <td>33.80</td>\n",
       "      <td>0.024378</td>\n",
       "      <td>0.081462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-163685</th>\n",
       "      <td>Integration of energy metabolism</td>\n",
       "      <td>0.044891</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.113509</td>\n",
       "      <td>0.34</td>\n",
       "      <td>34.00</td>\n",
       "      <td>0.028682</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-71182</th>\n",
       "      <td>Phenylalanine and tyrosine catabolism</td>\n",
       "      <td>0.069151</td>\n",
       "      <td>0.115252</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>54.17</td>\n",
       "      <td>0.015294</td>\n",
       "      <td>8.11</td>\n",
       "      <td>33.79</td>\n",
       "      <td>0.030229</td>\n",
       "      <td>0.054815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-351143</th>\n",
       "      <td>Agmatine biosynthesis</td>\n",
       "      <td>0.034558</td>\n",
       "      <td>0.100506</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>40.00</td>\n",
       "      <td>0.327631</td>\n",
       "      <td>1.69</td>\n",
       "      <td>33.80</td>\n",
       "      <td>0.030748</td>\n",
       "      <td>0.089895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-351200</th>\n",
       "      <td>Interconversion of polyamines</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.926579</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>11.11</td>\n",
       "      <td>0.904169</td>\n",
       "      <td>3.04</td>\n",
       "      <td>33.78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.953391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-2408517</th>\n",
       "      <td>SeMet incorporation into proteins</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.080412</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>33.33</td>\n",
       "      <td>0.415382</td>\n",
       "      <td>1.01</td>\n",
       "      <td>33.67</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.078244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-164378</th>\n",
       "      <td>PKA activation in glucagon signalling</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.080412</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>20.00</td>\n",
       "      <td>0.659417</td>\n",
       "      <td>1.69</td>\n",
       "      <td>33.80</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.097748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-77289</th>\n",
       "      <td>Mitochondrial Fatty Acid Beta-Oxidation</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.080412</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>12.50</td>\n",
       "      <td>0.866228</td>\n",
       "      <td>2.70</td>\n",
       "      <td>33.75</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.123507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-HSA-70221</th>\n",
       "      <td>Glycogen breakdown (glycogenolysis)</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.053633</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>30.00</td>\n",
       "      <td>0.541649</td>\n",
       "      <td>3.38</td>\n",
       "      <td>33.80</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.059520</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>167 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               pw_name  beer1/beer2 p-value  \\\n",
       "R-HSA-71240                      Tryptophan catabolism             0.025106   \n",
       "R-HSA-2024096                       HS-GAG degradation             0.018223   \n",
       "R-HSA-163685          Integration of energy metabolism             0.044891   \n",
       "R-HSA-71182      Phenylalanine and tyrosine catabolism             0.069151   \n",
       "R-HSA-351143                     Agmatine biosynthesis             0.034558   \n",
       "...                                                ...                  ...   \n",
       "R-HSA-351200             Interconversion of polyamines             1.000000   \n",
       "R-HSA-2408517        SeMet incorporation into proteins             1.000000   \n",
       "R-HSA-164378     PKA activation in glucagon signalling             1.000000   \n",
       "R-HSA-77289    Mitochondrial Fatty Acid Beta-Oxidation             1.000000   \n",
       "R-HSA-70221        Glycogen breakdown (glycogenolysis)             1.000000   \n",
       "\n",
       "               beer3/beer4 p-value  unq_pw_F  tot_ds_F  F_coverage        sf  \\\n",
       "R-HSA-71240               0.053509        27        14       51.85  0.020106   \n",
       "R-HSA-2024096             0.066147         5         1       20.00  0.659417   \n",
       "R-HSA-163685              1.000000         1         1      100.00  0.113509   \n",
       "R-HSA-71182               0.115252        24        13       54.17  0.015294   \n",
       "R-HSA-351143              0.100506         5         2       40.00  0.327631   \n",
       "...                            ...       ...       ...         ...       ...   \n",
       "R-HSA-351200              0.926579         9         1       11.11  0.904169   \n",
       "R-HSA-2408517             0.080412         3         1       33.33  0.415382   \n",
       "R-HSA-164378              0.080412         5         1       20.00  0.659417   \n",
       "R-HSA-77289               0.080412         8         1       12.50  0.866228   \n",
       "R-HSA-70221               0.053633        10         3       30.00  0.541649   \n",
       "\n",
       "               exp_F  Ex_Cov  COMPOUND beer1/beer2 comb_p  \\\n",
       "R-HSA-71240     9.12   33.78                     0.010104   \n",
       "R-HSA-2024096   1.69   33.80                     0.024378   \n",
       "R-HSA-163685    0.34   34.00                     0.028682   \n",
       "R-HSA-71182     8.11   33.79                     0.030229   \n",
       "R-HSA-351143    1.69   33.80                     0.030748   \n",
       "...              ...     ...                          ...   \n",
       "R-HSA-351200    3.04   33.78                     1.000000   \n",
       "R-HSA-2408517   1.01   33.67                     1.000000   \n",
       "R-HSA-164378    1.69   33.80                     1.000000   \n",
       "R-HSA-77289     2.70   33.75                     1.000000   \n",
       "R-HSA-70221     3.38   33.80                     1.000000   \n",
       "\n",
       "               COMPOUND beer3/beer4 comb_p  \n",
       "R-HSA-71240                       0.023696  \n",
       "R-HSA-2024096                     0.081462  \n",
       "R-HSA-163685                      1.000000  \n",
       "R-HSA-71182                       0.054815  \n",
       "R-HSA-351143                      0.089895  \n",
       "...                                    ...  \n",
       "R-HSA-351200                      0.953391  \n",
       "R-HSA-2408517                     0.078244  \n",
       "R-HSA-164378                      0.097748  \n",
       "R-HSA-77289                       0.123507  \n",
       "R-HSA-70221                       0.059520  \n",
       "\n",
       "[167 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df.sort_values('COMPOUND beer1/beer2 comb_p', ascending=True, inplace=True)\n",
    "pathway_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ORA Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:30:12.351 | DEBUG    | pals.feature_extraction:get_data:96 - Loading C:\\Users\\joewa\\Work\\git\\PALS\\pals\\data\\PiMP_KEGG.json.zip\n",
      "2019-12-18 13:30:12.406 | DEBUG    | pals.feature_extraction:__init__:46 - Mapping pathway to unique ids\n",
      "2019-12-18 13:30:12.412 | DEBUG    | pals.feature_extraction:__init__:60 - Creating dataset to pathway mapping\n",
      "2019-12-18 13:30:13.446 | DEBUG    | pals.feature_extraction:__init__:88 - Computing unique id counts\n"
     ]
    }
   ],
   "source": [
    "ds = DataSource(int_df, annotation_df, experimental_design, DATABASE_PIMP_KEGG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-12-18 13:30:13.526 | DEBUG    | pals.pathway_analysis:get_ora_df:50 - Calculating ORA\n",
      "2019-12-18 13:30:13.528 | DEBUG    | pals.pathway_analysis:_change_zero_peak_ints:385 - Setting the zero intensity values in the dataframe\n",
      "2019-12-18 13:30:19.806 | DEBUG    | pals.pathway_analysis:get_ora_df:109 - Correcting for multiple t-tests\n",
      "2019-12-18 13:30:19.812 | DEBUG    | pals.pathway_analysis:_calculate_coverage_df:471 - Calculating dataset formula coverage\n"
     ]
    }
   ],
   "source": [
    "pals = PALS(ds, min_replace=5000)\n",
    "pathway_df = pals.get_ora_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw_name</th>\n",
       "      <th>beer1/beer2 p-value</th>\n",
       "      <th>beer3/beer4 p-value</th>\n",
       "      <th>PiMP_KEGG beer1/beer2 comb_p</th>\n",
       "      <th>PiMP_KEGG beer3/beer4 comb_p</th>\n",
       "      <th>unq_pw_F</th>\n",
       "      <th>tot_ds_F</th>\n",
       "      <th>F_coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mapids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>map00350</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>3.621191e-16</td>\n",
       "      <td>1.681454e-16</td>\n",
       "      <td>8.183892e-14</td>\n",
       "      <td>3.800087e-14</td>\n",
       "      <td>53</td>\n",
       "      <td>39</td>\n",
       "      <td>73.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00330</th>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>1.775486e-12</td>\n",
       "      <td>1.686603e-12</td>\n",
       "      <td>2.006299e-10</td>\n",
       "      <td>1.905861e-10</td>\n",
       "      <td>79</td>\n",
       "      <td>50</td>\n",
       "      <td>63.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00290</th>\n",
       "      <td>Valine, leucine and isoleucine biosynthesis</td>\n",
       "      <td>9.591121e-12</td>\n",
       "      <td>2.028069e-11</td>\n",
       "      <td>7.225312e-10</td>\n",
       "      <td>1.527812e-09</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>94.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00400</th>\n",
       "      <td>Phenylalanine, tyrosine and tryptophan biosynt...</td>\n",
       "      <td>2.824918e-10</td>\n",
       "      <td>7.042130e-10</td>\n",
       "      <td>1.596078e-08</td>\n",
       "      <td>3.978804e-08</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>73.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00660</th>\n",
       "      <td>C5-Branched dibasic acid metabolism</td>\n",
       "      <td>5.693949e-09</td>\n",
       "      <td>1.159169e-08</td>\n",
       "      <td>1.838332e-07</td>\n",
       "      <td>2.619722e-07</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>76.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00254</th>\n",
       "      <td>Aflatoxin biosynthesis</td>\n",
       "      <td>9.905000e-01</td>\n",
       "      <td>9.925788e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "      <td>12.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00195</th>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.696482e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.839475e-01</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00981</th>\n",
       "      <td>Insect hormone biosynthesis</td>\n",
       "      <td>9.829573e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>9.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00942</th>\n",
       "      <td>Anthocyanin biosynthesis</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>map00351</th>\n",
       "      <td>DDT degradation</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.831430e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>226 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    pw_name  \\\n",
       "mapids                                                        \n",
       "map00350                                Tyrosine metabolism   \n",
       "map00330                    Arginine and proline metabolism   \n",
       "map00290        Valine, leucine and isoleucine biosynthesis   \n",
       "map00400  Phenylalanine, tyrosine and tryptophan biosynt...   \n",
       "map00660                C5-Branched dibasic acid metabolism   \n",
       "...                                                     ...   \n",
       "map00254                             Aflatoxin biosynthesis   \n",
       "map00195                                     Photosynthesis   \n",
       "map00981                        Insect hormone biosynthesis   \n",
       "map00942                           Anthocyanin biosynthesis   \n",
       "map00351                                    DDT degradation   \n",
       "\n",
       "          beer1/beer2 p-value  beer3/beer4 p-value  \\\n",
       "mapids                                               \n",
       "map00350         3.621191e-16         1.681454e-16   \n",
       "map00330         1.775486e-12         1.686603e-12   \n",
       "map00290         9.591121e-12         2.028069e-11   \n",
       "map00400         2.824918e-10         7.042130e-10   \n",
       "map00660         5.693949e-09         1.159169e-08   \n",
       "...                       ...                  ...   \n",
       "map00254         9.905000e-01         9.925788e-01   \n",
       "map00195         1.000000e+00         8.696482e-01   \n",
       "map00981         9.829573e-01         1.000000e+00   \n",
       "map00942         1.000000e+00         1.000000e+00   \n",
       "map00351         1.000000e+00         9.831430e-01   \n",
       "\n",
       "          PiMP_KEGG beer1/beer2 comb_p  PiMP_KEGG beer3/beer4 comb_p  \\\n",
       "mapids                                                                 \n",
       "map00350                  8.183892e-14                  3.800087e-14   \n",
       "map00330                  2.006299e-10                  1.905861e-10   \n",
       "map00290                  7.225312e-10                  1.527812e-09   \n",
       "map00400                  1.596078e-08                  3.978804e-08   \n",
       "map00660                  1.838332e-07                  2.619722e-07   \n",
       "...                                ...                           ...   \n",
       "map00254                  1.000000e+00                  1.000000e+00   \n",
       "map00195                  1.000000e+00                  9.839475e-01   \n",
       "map00981                  1.000000e+00                  1.000000e+00   \n",
       "map00942                  1.000000e+00                  1.000000e+00   \n",
       "map00351                  1.000000e+00                  1.000000e+00   \n",
       "\n",
       "          unq_pw_F  tot_ds_F  F_coverage  \n",
       "mapids                                    \n",
       "map00350        53        39       73.58  \n",
       "map00330        79        50       63.29  \n",
       "map00290        17        16       94.12  \n",
       "map00400        30        22       73.33  \n",
       "map00660        21        16       76.19  \n",
       "...            ...       ...         ...  \n",
       "map00254        24         3       12.50  \n",
       "map00195        10         1       10.00  \n",
       "map00981        21         2        9.52  \n",
       "map00942        50         1        2.00  \n",
       "map00351        20         2       10.00  \n",
       "\n",
       "[226 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df.sort_values('PiMP_KEGG beer1/beer2 comb_p', ascending=True, inplace=True)\n",
    "pathway_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Django Shell-Plus",
   "language": "python",
   "name": "django_extensions"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
